content/5_Plat/RURQ.mdx

```diff
@@ -152,135 +152,135 @@ Given the mechanics behind the `pushdown` function, all you need is a regular
 range-sum segment tree to solve the question.
 
 ```cpp
-
-#include <iostream>
-using ll = long long;
+#include <bits/stdc++.h>
 using namespace std;
+using ll = long long;
 
-const int maxN = 2e5 + 5;
-
-int N, Q;
-int a[maxN];
+/**
+ * Represents the type of lazy update being done.
+ * NONE = if there's no lazy update to be performed.
+ */
+enum QueryType { ADD, SET, NONE };
 
-struct node {
-	ll val;
-	ll lzAdd;
-	ll lzSet;
-	node(){};
-} tree[maxN << 2];
+struct Query {
+	QueryType type = NONE;
+	ll val = 0;
+};
 
-#define lc p << 1
-#define rc (p << 1) + 1
+/**
+ * T = tree node type, which will be long long
+ * U = update type, which will be the Query struct
+ */
+template <class T, class U> class LazySegtree {
+  private:
+	const int sz;
+	vector<T> tree;  // tree[i] = sum of this node's range
+	vector<U> lazy;  // lazy[i] = lazy update for the range
+
+	/** @return result of joining two tree nodes together */
+	inline T comb(T a, T b) { return a + b; }
+
+	/** builds the segtree nodes */
+	void build(int v, int l, int r, const vector<T> &a) {
+		if (l == r) {
+			tree[v] = a[l];
+		} else {
+			int m = (l + r) / 2;
+			build(2 * v, l, m, a);
+			build(2 * v + 1, m + 1, r, a);
+			tree[v] = comb(tree[2 * v], tree[2 * v + 1]);
+		}
+	}
 
-inline void pushup(int p) {
-	tree[p].val = tree[lc].val + tree[rc].val;
-	return;
-}
+	/** applies lazy update to t[v], places update at lz[v] */
+	void apply(int v, int len, U x) {
+		if (x.type == ADD) {
+			// if lazy[v]'s type is NONE or ADD, then we add to the range
+			// otherwise, we add to our lazy set value
+			if (lazy[v].type != SET) {
+				lazy[v] = Query{ADD, lazy[v].val + x.val};
+			} else {
+				lazy[v] = Query{SET, lazy[v].val + x.val};
+			}
+			tree[v] += x.val * len;
+		} else {
```

Comment:
```suggestion
		} else if (query.type == SET) {
```

Reply:
idk its implied alr but why not

Reply:
it probably isn't called with none but
smth smth explicit > implicit